Design doc for Remote Reference

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30066 Design doc for Remote Reference**
* #30160 Stop showing .api as an API path component in RPC docs

This commit adds design reasoning and walks through four scenarios
for RRef.

Differential Revision: [D18595094](https://our.internmc.facebook.com/intern/diff/D18595094)

mrshenli committed:
Design doc for Remote Reference

This commit adds design reasoning and walks through four scenarios
for RRef.

[ghstack-poisoned]

mrshenli commented:
![rref_screenshot](https://user-images.githubusercontent.com/16999635/69114667-05428e80-0a54-11ea-8894-dd1cbe6d96b7.png)

mrshenli committed:
Update on "Design doc for Remote Reference"


This commit adds design reasoning and walks through four scenarios
for RRef.

[ghstack-poisoned]

mrshenli committed:
Update on "Design doc for Remote Reference"


This commit adds design reasoning and walks through four scenarios
for RRef.

Differential Revision: [D18595094](https://our.internmc.facebook.com/intern/diff/D18595094)

[ghstack-poisoned]

rohan-varma commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+Background
+^^^^^^^^^^
+
+RRef stands for Remote REFerence. It is a reference of an object which locates
```
change "which locates" to "which is located" for additional clarity?

rohan-varma commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+on the local or a remote worker, and transparently handles reference counting
+under the hood. Conceptually, it can be considered as a distributed shared
+pointer. Applications can create an RRef by calling
+`torch.distributed.rpc.remote`. Each RRef is owned by the callee worker of the
```
It might be helpful to link to all these functions. You can do it with `func: torch.distributed.rpc.remote`.

rohan-varma commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+retrieve the unique ``OwnerRRef`` instance using the globally unique ``RRefId``.
+A ``UserRRef`` will be created when it is used as an argument or return value in
+``rpc`` or ``remote`` invocation, and the owner will be notified according to
+update the reference count. An ``OwnerRRef`` and its data will be deleted when
```
Nit, are these classes clickable links in the docs?

mrshenli replied:
No, they are not currently, let me make changes, thanks!

rohan-varma commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+3) Creating a new ``UserRRef`` owned by another worker.
+
+
+Case 1 is the simplest where the owner initiates the fork, and hence it can
```
Describe what "initiates the fork" means? I'm assuming this means they call `rpc.remote` or something similar.

mrshenli replied:
Yes, let me edit

rohan-varma commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+B that involves an RRef, we call the RRef on A the parent RRef and the RRef on B
+the child RRef.
+
+**G2. Parent RRef will be deleted until the child RRef is confirmed by the
```
Is it parent RRef will NOT be deleted?

jlin27 commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+Assumptions
+^^^^^^^^^^^
+
+RRef protocol is design with the following assumptions.
```
"is design[ed]"

rohan-varma commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+owner.**
+
+In cases 2 and 3, it is possible that the owner only partially knows the RRef
+fork graph or not even knowing it at all. For example, an RRef could be
```
"it is possible that the owner has only partial or no knowledge  at all about the RRef fork graph" reads a bit better imo

rohan-varma commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+In cases 2 and 3, it is possible that the owner only partially knows the RRef
+fork graph or not even knowing it at all. For example, an RRef could be
+constructed on a user, and before the owner receives any RPC call, the
+creator user might have already shared the RRef with other users, and those
```
Nit: it might be helpful to add a code scenario of where this could happen? Just a couple of lines that create an RRef fork tree by sharing RRefs. This might make it easier for someone to follow - it can be done as a follow up task.

rohan-varma replied:
ignore this, i just read a bit more and saw the code examples

rohan-varma commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+creator user might have already shared the RRef with other users, and those
+users could further share the RRef. One invariant is that the fork graph of
+any RRef is always a tree, because forking an RRef always
+creates a new ``UserRRef`` instance on the callee (except the callee is the
```
except IF the callee is the owner is a bit clearer

jlin27 commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+
+On the owner, when receiving the ``rpc.remote`` call, it will create the
+``OwnerRRef``, and returns an ACK to acknowledge ``{100, 1}``. Only
+after receiving this ACK, can A delete it's ``UserRRef``. This involves both
```
"it's" -> "its"

jlin27 commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+**G1** is implemented by sending out a delete message in ``UserRRef``
+destructor. To provide **G2**, the parent ``UserRRef`` is put into a context
+whenever it is forked, indexed by the new ``ForkId``. The parent ``UserRRef`` is
+only removed from the context when it receives an ACK from the child, and the
```
What's ACK? Considering linking to relevant doc for users who aren't familiar or add a short explanation.

zhaojuanmao replied:
nit: expand ACK

jlin27 commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+it.
+
+On the owner, when receiving the ``rpc.remote`` call, it will create the
+``OwnerRRef``, and returns an ACK to acknowledge ``{100, 1}``. Only
```
Why the values {100,1}? Is it RRefId 100 and ForkId 1?

mrshenli replied:
Yes, let me add that

jlin27 commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+The figure above shows the message flow. Note that when the ``OwnerRRef`` exits
+scope after the rpc_async call, it will not be deleted, because internally
+there is a map to hold it alive if there is any known forks, in which case is
+``UserRRef {100, 2}``. (**G2**)
```
Why {100, 2} ?

mrshenli replied:
shuold be {100, 1}

jlin27 commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+
+When C receives the child ``UserRRef`` from A, it sends out a fork request to
+the owner B. Later, when the B confirms the ``UserRRef`` on C, C will perform
+two actions in parallel: 1) send out the child ACK to A and 2) run the UDF.
```
What's UDF? Considering linking to relevant doc for users who aren't familiar or add a short explanation.

zhaojuanmao replied:
expand UDF

rohan-varma commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+final delete message will only be sent out when:
+
+- B acknowledges ``UserRRef {100, 1}`` (G2), and
+- Python GC agrees to delete the local ``UserRRef`` instance.
```
Clarify that this happens when the rref is no longer in scope and it is eligible for garbage collection?

rohan-varma reviewed (commented):
Looks good % minor comments.

zhaojuanmao commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+  partition is beyond the scope. When those incidents occur, the application
+  may take down all workers, revert to the previous checkpoint, and resume
+  training.
+- **Non-idempotent UDFs**: We assume UDFs are not idempotent and therefore
```
nit: explain  UDF

zhaojuanmao commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+
+The goal of the protocol is to delete an ``OwnerRRef`` at an appropriate time.
+The right time to delete an ``OwnerRRef`` is when there are no living
+``UserRRef`` instances and Python GC also agrees to delete the ``OwnerRRef``
```
nit: expand GC

zhaojuanmao commented on docs/source/notes/rref.rst:
```diff
@@ -0,0 +1,304 @@
+destructor. To provide **G2**, the parent ``UserRRef`` is put into a context
+whenever it is forked, indexed by the new ``ForkId``. The parent ``UserRRef`` is
+only removed from the context when it receives an ACK from the child, and the
+child will only send out the ACK when it is confirmed by the owner.
```
nit: expand ACK

zhaojuanmao reviewed (commented):
do you want to add some explanation for create RREF to itself, it will return an owner rref? also RREF(t) API ?

mrshenli committed:
Update on "Design doc for Remote Reference"


This commit adds design reasoning and walks through four scenarios
for RRef.

Differential Revision: [D18595094](https://our.internmc.facebook.com/intern/diff/D18595094)

[ghstack-poisoned]

mrshenli committed:
Update on "Design doc for Remote Reference"


This commit adds design reasoning and walks through four scenarios
for RRef.

Differential Revision: [D18595094](https://our.internmc.facebook.com/intern/diff/D18595094)

[ghstack-poisoned]

mrshenli committed:
Update on "Design doc for Remote Reference"


This commit adds design reasoning and walks through four scenarios
for RRef.

Differential Revision: [D18595094](https://our.internmc.facebook.com/intern/diff/D18595094)

[ghstack-poisoned]

mrshenli commented:
> do you want to add some explanation for create RREF to itself, it will return an owner rref? also RREF(t) API ?

Will do this in followup PRs
